abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves all key-value pairs from the header map `map_id`.
+
+`return_pairs` is [serialized].
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `map_id`.
+- `INVALID_MEMORY_ACCESS` when `return_pairs_data` and/or
+  `return_pairs_size` point to invalid memory address.
+
+
+### `proxy_set_header_map_pairs`
+
+* params:
+  - `i32 (`[`proxy_header_map_type_t`]`) map_id`
+  - `i32 (const uint8_t *) serialized_pairs_data`
+  - `i32 (size_t) serialized_pairs_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Sets all key-value pairs in the header map `map_id` to `pairs`.
+
+`pairs` must be [serialized].
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `map_id`.
+- `INVALID_MEMORY_ACCESS` when `pairs_data` and/or
+  `pairs_size` point to invalid memory address.
+
+
+### `proxy_get_header_map_value`
+
+* params:
+  - `i32 (`[`proxy_header_map_type_t`]`) map_id`
+  - `i32 (const char *) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (uint8_t **) return_value_data`
+  - `i32 (size_t *) return_value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Retrieves value of the header `key` from the header map `map_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `map_id`.
+- `NOT_FOUND` when the requested `key` was not found.
+- `INVALID_MEMORY_ACCESS` when `key_data`, `key_size`,
+  `return_value_data` and/or `return_value_size` point to
+  invalid memory address.
+
+
+### `proxy_add_header_map_value`
+
+* params:
+  - `i32 (`[`proxy_header_map_type_t`]`) map_id`
+  - `i32 (const char *) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const uint8_t *) value_data`
+  - `i32 (size_t) value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Adds `key` with `value` to the header map `map_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `map_id`.
+- `INVALID_MEMORY_ACCESS` when `key_data`, `key_size`, `value_data`
+  and/or `value_size` point to invalid memory address.
+
+
+### `proxy_replace_header_map_value`
+
+* params:
+  - `i32 (`[`proxy_header_map_type_t`]`) map_id`
+  - `i32 (const char *) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (const uint8_t *) value_data`
+  - `i32 (size_t) value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Adds or replaces `key`'s value with `value` in the header map `map_id`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `map_id`.
+- `INVALID_MEMORY_ACCESS` when `key_data`, `key_size`, `value_data`
+  and/or `value_size` point to invalid memory address.
+
+
+### `proxy_remove_header_map_value`
+
+* params:
+  - `i32 (`[`proxy_header_map_type_t`]`) map_id`
+  - `i32 (const char *) key_data`
+  - `i32 (size_t) key_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Removes the `key` from the header map `map_id`.
+
+Returned `status` value is:
```

Comment:
It does not, calling `proxy_remove_header_map_value(..., "foo", 3)` will return `OK`, since it reaches the desired state and the key named "foo" doesn't exist after this call returns.

I could be convinced that returning `NOT_FOUND` (at least at the ABI layer, SDKs can do whatever feels more idiomatic) would be better, but that's not what's currently implemented.